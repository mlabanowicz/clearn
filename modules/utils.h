
#ifndef UTILS_H__
#define UTILS_H__

#include <stddef.h> /* size_t */
#include <stdio.h>  /* FILE   */

#ifdef __cplusplus
extern "C" {
#endif

/* Print some information about build date/time to specified FILE stream */
void UTILS_BuildInfo(FILE * stream);

/* Funkcja zaladuje wszystkie dane ze strumienia do bufora zaalokowanego w pamieci HEAP
 * Uzytkownik na koncu uzywania bufora powinien zwolnic ten bufor za pomoca free
 * Przyklad uzycia:
 * {
 *   void * buffer = NULL;
 *   size_t length = 0;
 *
 *   buffer = UTILS_LoadFile(stdin, &length);
 *
 *   if (buffer) {
 *
 *     printf("OK, zaladowano dane ze strumienia\n");
 *
 *     printf("Adres z danymi: %p\n", buffer);
 *     printf("Rozmiar danych: %lu bajtow\n", (unsigned long)length);
 *
 *     free(buffer);
 *
 *   } else {
 *
 *     fprintf(stderr, "Jakis blad, nie zaladowano danych !\n");
 *
 *   }
 * }
 */
void * UTILS_LoadFile(FILE * stream, size_t * size_ptr);

/* Funkcja wyswietli zawartosc danych o adresie data i rozmiarze size
 * w postacji zrzutu pamieci z linia o 16 bajtach.
 *
 * 1 kolumna: offset o rozmiarze 8 znakow,
 *            liczony od poczatku podanego adresu danych
 * 2 kolumna: 16 bajtow w systemie szesnastkowym
 * 3 kolumna: 16 bajtow w systemie znakowym, jezeli jakis znak
 *            znaki spoza zakresu 32..126 wyswietlane w postaci kropki.
 *
 * Gdy ostatnia linia nie jest pelna (size nie jest wyalognowany do 16),
 * wtedy w miejsca tych bajtow nalezy wpisac spacje.
 *
 * | <offset> | aa aa aa aa aa aa aa aa  bb bb bb bb bb bb bb bb | cccccccccccccccc |
 */
void UTILS_DumpData(FILE * output, void const * data, size_t size);

/* Funckja zwroci liczbe bitow ustawionych
 * na podanym obszarze pamieci
 *
 * Przykladowo, gdy pamiec wyglada tak (hex):
 * data = 00 01 02 03, size = 4 (bajty)
 * zwroci: 0 + 1 + 1 + 2 = 4 - bo tyle bitow jest ustawionych
 */
size_t UTILS_BitsCount(void const * data, size_t size);

#ifdef __cplusplus
}
#endif

#endif /* UTILS_H__ */
