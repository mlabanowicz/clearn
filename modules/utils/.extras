
build()
{
  (cd bitscnt && ./generuj_tablice.sh)
}

clean()
{
  (cd bitscnt && ./clean.sh)
}

