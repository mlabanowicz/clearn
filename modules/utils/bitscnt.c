
#include <utils.h>

#ifdef __cplusplus
extern "C" {
#endif

size_t UTILS_BitsCount(void const *data, size_t size)
{
  size_t counter = 0;

  unsigned char const *ptr = (unsigned char const *)data;

  size_t i;

  static unsigned char const array [] =
  {
    #include "bitscnt/tablica.txt"
  };

  for (i = 0; i < size; i++)
  {
      counter += array[*ptr++];
  }

  return counter;
}

#ifdef __cplusplus
}
#endif
