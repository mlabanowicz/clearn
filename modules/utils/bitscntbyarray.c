#include <utils.h>
#include <stdlib.h>
#include <limits.h>
#include <stdio.h>
#ifdef __cplusplus
extern "C"
{
#endif

    size_t UTILS_BitsCountByArray(void const *data, size_t size)
    {
        size_t counter = 0;
        unsigned char *ptr = (unsigned char *)data;
        size_t i;
        unsigned char j;
        unsigned char * ptr2 = malloc(CHAR_BIT * 256),* array = ptr2;
        for(i=0;i < 256;i++){
            *ptr2 = 0;
            for (j = 0; j < CHAR_BIT; j++)
            {
                  *ptr2 += (i >> j) & 1;
            }
            ptr2++;
        }
        for (i = 0; i < size; i++)
        {
            counter += array[(short)*ptr];
            ptr++;
        }
        free(array);
        return counter;
    }

#ifdef __cplusplus
}
#endif
