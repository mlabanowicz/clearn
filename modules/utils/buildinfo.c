
#include <utils.h>

#ifdef __cplusplus
extern "C" {
#endif

void UTILS_BuildInfo(FILE * stream) {
  fputs("Program build date: " __DATE__ " / " __TIME__ "\n", stream);
}

#ifdef __cplusplus
}
#endif
