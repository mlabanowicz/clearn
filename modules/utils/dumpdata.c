#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#include <utils.h>

#ifdef __cplusplus
extern "C"
{
#endif

    void UTILS_DumpData(FILE *output, void const *data, size_t size)
    {
        size_t i;
        int j = 0;
        /*char const * column = "| ";*/
        char *ptr = (char *)data;
        for (i = 0; i < (size / 16) + 1 && i * 16 <= size; i++)
        {
            fprintf(output, " | ");
            for (j = 0; j < 16; j++)
            {
                if (j == 0)
                    fprintf(output, "%08lX | ", (unsigned long)(i * 16) + j);
                if ((i * 16) + j >= size)
                {
                    fprintf(output, "   ");
                }
                else
                {
                    fprintf(output, "%02X ", ptr[(i * 16) + j]);
                    
                }
                if (j == 7){
                        fprintf(output, " ");
                    }
            }
            fprintf(output, "| ");
            for (j = 0; j < 16; j++)
            {
                if ((i * 16) + j >= size)
                {
                    fprintf(output, " ");
                }
                else
                {
                    fprintf(output, "%c", (ptr[(i * 16) + j] >= ' ' && ptr[(i * 16) + j] < SCHAR_MAX) ? ptr[(i * 16) + j] : '.');
                }
            }
            fprintf(output, " |\n");
        }
    }

#ifdef __cplusplus
}
#endif
