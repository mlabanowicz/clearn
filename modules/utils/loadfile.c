#include <stdlib.h>
#include <stdio.h>
#include <limits.h>

#include <utils.h>

#ifdef __cplusplus
extern "C"
{
#endif
    void *UTILS_LoadFile(FILE *stream, size_t *size_ptr)
    {
        char   * buffer = NULL; /* Bufor wynikowy                  */
        size_t   length = 0;    /* Liczba bajtow uzytych w buforze */

        /* Odczytywanie danych ze strumienia, o ktorym nic nie wiemy,
         * mozna zrealizowac za pomoca fgetc */
        int c;

        while ((c = fgetc(stream)) != EOF) {

          /* OK, w strumieniu jest dodatkowy bajt, dodajmy go do wyjsciowego bufora: */

          void * ptr = realloc(buffer, length + 1);

          if (!ptr) {
            fprintf(stderr, "LOG:blad realokacji HEAP\n");
            free(buffer);
            exit(EXIT_FAILURE);
          }

          buffer = (char *)ptr; /* buffer ma teraz nowy adres, i jest wiekszy o 1 bajt */

          /* wrzucmy ten bajt na koniec tego bufora i pozniej
           * zwiekszmy zmienna 'length' przechowujaca jego rozmiar */
          buffer[length++] = (unsigned char)c;

        }

        /* *size_ptr jest to parametr, ktory powinien byc wypelniony przez ta funkcje */
        *size_ptr = length;

        return buffer;
    }

#ifdef __cplusplus
}
#endif
