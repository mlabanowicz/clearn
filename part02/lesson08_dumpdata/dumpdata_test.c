
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>

#include <utils.h>

#ifdef __cplusplus
extern "C" {
#endif

int main(void)
{
  size_t   size = 0;
  void   * data = UTILS_LoadFile(stdin, &size);

  UTILS_DumpData(stdout, data, size);

  free(data);

  return EXIT_SUCCESS;
}

#ifdef __cplusplus
}
#endif
