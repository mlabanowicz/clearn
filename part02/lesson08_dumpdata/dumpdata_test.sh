#!/bin/bash -ex

# sciezka do katalogu modules
MODULES=../../modules

# zbudowanie biblioteki utils.a
${MODULES}/build.sh utils

# zaladowanie opcji kompilacji
. ${MODULES}/opts.sh

# kompilacja programu z wykorzystaniem biblioteki
gcc ${opts} ${opts_rel} dumpdata_test.c -I${MODULES} ${MODULES}/utils.a -o dumpdata_test.exe

# uruchomienie/test programu (jako wejscie stdin podano ten skrypt)
./dumpdata_test.exe < "${0}"
