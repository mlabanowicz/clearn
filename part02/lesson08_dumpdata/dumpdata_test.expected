 | 00000000 | 23 21 2F 62 69 6E 2F 62  61 73 68 20 2D 65 78 0A | #!/bin/bash -ex. |
 | 00000010 | 0A 23 20 73 63 69 65 7A  6B 61 20 64 6F 20 6B 61 | .# sciezka do ka |
 | 00000020 | 74 61 6C 6F 67 75 20 6D  6F 64 75 6C 65 73 0A 4D | talogu modules.M |
 | 00000030 | 4F 44 55 4C 45 53 3D 2E  2E 2F 2E 2E 2F 6D 6F 64 | ODULES=../../mod |
 | 00000040 | 75 6C 65 73 0A 0A 23 20  7A 62 75 64 6F 77 61 6E | ules..# zbudowan |
 | 00000050 | 69 65 20 62 69 62 6C 69  6F 74 65 6B 69 20 75 74 | ie biblioteki ut |
 | 00000060 | 69 6C 73 2E 61 0A 24 7B  4D 4F 44 55 4C 45 53 7D | ils.a.${MODULES} |
 | 00000070 | 2F 62 75 69 6C 64 2E 73  68 20 75 74 69 6C 73 0A | /build.sh utils. |
 | 00000080 | 0A 23 20 7A 61 6C 61 64  6F 77 61 6E 69 65 20 6F | .# zaladowanie o |
 | 00000090 | 70 63 6A 69 20 6B 6F 6D  70 69 6C 61 63 6A 69 0A | pcji kompilacji. |
 | 000000A0 | 2E 20 24 7B 4D 4F 44 55  4C 45 53 7D 2F 6F 70 74 | . ${MODULES}/opt |
 | 000000B0 | 73 2E 73 68 0A 0A 23 20  6B 6F 6D 70 69 6C 61 63 | s.sh..# kompilac |
 | 000000C0 | 6A 61 20 70 72 6F 67 72  61 6D 75 20 7A 20 77 79 | ja programu z wy |
 | 000000D0 | 6B 6F 72 7A 79 73 74 61  6E 69 65 6D 20 62 69 62 | korzystaniem bib |
 | 000000E0 | 6C 69 6F 74 65 6B 69 0A  67 63 63 20 24 7B 6F 70 | lioteki.gcc ${op |
 | 000000F0 | 74 73 7D 20 24 7B 6F 70  74 73 5F 72 65 6C 7D 20 | ts} ${opts_rel}  |
 | 00000100 | 64 75 6D 70 64 61 74 61  5F 74 65 73 74 2E 63 20 | dumpdata_test.c  |
 | 00000110 | 2D 49 24 7B 4D 4F 44 55  4C 45 53 7D 20 24 7B 4D | -I${MODULES} ${M |
 | 00000120 | 4F 44 55 4C 45 53 7D 2F  75 74 69 6C 73 2E 61 20 | ODULES}/utils.a  |
 | 00000130 | 2D 6F 20 64 75 6D 70 64  61 74 61 5F 74 65 73 74 | -o dumpdata_test |
 | 00000140 | 2E 65 78 65 0A 0A 23 20  75 72 75 63 68 6F 6D 69 | .exe..# uruchomi |
 | 00000150 | 65 6E 69 65 2F 74 65 73  74 20 70 72 6F 67 72 61 | enie/test progra |
 | 00000160 | 6D 75 20 28 6A 61 6B 6F  20 77 65 6A 73 63 69 65 | mu (jako wejscie |
 | 00000170 | 20 73 74 64 69 6E 20 70  6F 64 61 6E 6F 20 74 65 |  stdin podano te |
 | 00000180 | 6E 20 73 6B 72 79 70 74  29 0A 2E 2F 64 75 6D 70 | n skrypt)../dump |
 | 00000190 | 64 61 74 61 5F 74 65 73  74 2E 65 78 65 20 3C 20 | data_test.exe <  |
 | 000001A0 | 22 24 7B 30 7D 22 0A                             | "${0}".          |
